porting_guide_core_2.19.rst: add clatifications on the required non-string key conversion (#2560)
ansible/ansible-documentation: docs/docsite/rst/porting_guides/porting_guide_core_2.19.rst

Special handling in plugins is no longer required to access the contents of these values.

No implicit conversion of non-string keys in returned dictionaries
------------------------------------------------------------------

> Previously, ``ansible-core`` relied on the JSON dumping to implicitly convert ``non-string dictionary keys`` to strings when modules returned data.

Reviewer: ```suggestion
In previous versions, ``ansible-core`` relied on JSON dumping to implicitly convert ``non-string dictionary keys`` to strings when modules returned data.
```

> Starting with this release, modules must explicitly convert any non-string keys to strings before passing dictionaries to the ``AnsibleModule.exit_json()`` method of ``ansible-core``.

Reviewer: Are there any specific docs on how to implement the conversion? It would probably be good to link that here.

Author: SGTM, added one, PTAL

Author: @oraNod added more, PTAL

> If you encounter ``"[ERROR]: Task failed: Module failed: Key of type '<NON-STRING>' is not JSON serializable by the 'module_legacy_m2c' profile.``, it indicates that the module being used in the task does not perform the required key conversion.

Reviewer: ```suggestion
If you encounter ``"[ERROR]: Task failed: Module failed: Key of type '<NON-STRING>' is not JSON serializable by the 'module_legacy_m2c' profile.``, it indicates that the module that is used in the task does not perform the required key conversion.
```

> Please report this behavior in the appropriate module's repository and reference this section for context.

Reviewer: ```suggestion
Please report this behavior by opening an issue in the appropriate repository for any module that should convert keys to string.
Include a link to this section of the porting guide in the issue for context.
```

Author: Actually, users don't know which module should do it

Special handling in plugins is no longer required to access the contents of these values.

> No implicit conversion of non-string dict keys
> ------------------------------------------------------------------

Reviewer 2: ```suggestion
No implicit conversion of non-string dict keys
----------------------------------------------
```

Author: @mattclay committed, thanks, PTAL

> In previous versions, ``ansible-core`` relied on Python's ``json.dumps`` to implicitly convert ``int``, ``float``, ``bool`` and `None`` dictionary keys to strings in various scenarios, including returning of module results.

Author: ```suggestion
In previous versions, ``ansible-core`` relied on Python's ``json.dumps`` to implicitly convert ``int``, ``float``, ``bool`` and ``None`` dictionary keys to strings in various scenarios, including returning of module results.
```

Special handling in plugins is no longer required to access the contents of these values.

> No implicit conversion of non-string keys in returned dictionaries

Reviewer 2: This technically applies to any JSON dumping using an Ansible profile-based JSON serializer, not just data returned from modules -- that just happens to be the most common case.

```suggestion
No implicit conversion of non-string dict keys
```

------------------------------------------------------------------

> In previous versions, ``ansible-core`` relied on JSON dumping to implicitly convert ``non-string dictionary keys`` to strings when modules returned data.

Reviewer 2: ```suggestion
In previous versions, ``ansible-core`` relied on Python's ``json.dumps`` to implicitly convert ``int``, ``float``, ``bool`` and `None`` dictionary keys to strings in various scenarios, including returning of module results.

For example, a module was allowed to contain the following code:

.. code-block:: python

    oid = 123
    d = {oid: "value"}
    module.exit_json(return_value=d)

Starting with this release, modules must explicitly convert any non-string keys to strings (for example, by using the ``str()`` Python function) before passing dictionaries to the ``AnsibleModule.exit_json()`` method of ``ansible-core``. The above code must be changed as follows:

.. code-block:: python

    oid = 123
    d = {str(oid): "value"}
    module.exit_json(return_value=d)

If you encounter ``"[ERROR]: Task failed: Module failed: Key of type '<NON-STRING>' is not JSON serializable by the 'module_legacy_m2c' profile.``, it indicates that the module that is used in the task does not perform the required key conversion.

> 
> Please report this behavior in the appropriate module's repository and reference this section for context.

Reviewer 2: I'd omit this statement from the porting guide.

There's nothing specific about this particular scenario that would make this necessary.

```suggestion
```
